components/eam/src/physics/cam/micro_p3.F90

```diff
@@ -2878,64 +2878,70 @@ end subroutine ice_supersat_conservation
 
-   implicit none
+  implicit none
 
-   real(rtype), intent(in) :: pres
-   real(rtype), intent(in) :: t_atm
-   real(rtype), intent(in) :: qv
-   real(rtype), intent(in) :: latent_heat_sublim,latent_heat_vapor
-   real(rtype), intent(in) :: dt
-   real(rtype), intent(in) :: qidep,qinuc
-   real(rtype), intent(inout) :: qi2qv_sublim_tend,qr2qv_evap_tend
-   
-   real(rtype) :: qv_sinks, qv_sources, qv_endstep, T_endstep, qsl, A, frac
-
-   qv_sinks   = qidep + qinuc
-   qv_sources = qi2qv_sublim_tend + qr2qv_evap_tend
-   
-   !Actual qv after micro step
-   qv_endstep = qv - qv_sinks*dt + qv_sources*dt 
-   ! ... corresponding temperature
-   T_endstep = t_atm + ( (qv_sinks-qi2qv_sublim_tend)*latent_heat_sublim*inv_cp &
-               - qr2qv_evap_tend*latent_heat_vapor*inv_cp )*dt
-   
-   !qv we would have at end of step if we were saturated with respect to liquid
-   qsl = qv_sat(T_endstep,pres,0)
-   
-   ! The balance we seek is:
-   ! qv-qv_sinks*dt+qv_sources*frac*dt=qsl+dqsl_dT*(T correction due to conservation)
-   ! where the T correction for conservation is:
-   ! dt*[latent_heat_sublim/cp*(qi2qv_sublim_tend-frac*qi2qv_sublim_tend)
-   !     +latent_heat_vapor /cp*(qr2qv_evap_tend  -frac*qr2qv_evap_tend)]
-   ! =(1-frac)*dt/cp*(latent_heat_sublim*qi2qv_sublim_tend + latent_heat_vap*qr2qv_evap_tend).
-   ! Note T correction is positive because frac *reduces* evaporative cooling. Note as well that
-   ! dqsl_dt comes from linearization of qsl around the end-of-step T computed before temperature
-   ! correction. dqsl_dt should be computed with respect to *liquid* even though frac also adjusts
-   ! sublimation because we want to be saturated with respect to liquid at the end of the step.
-   ! dqsl_dt=Latent_heat_vapor*qsl/rv*T^2 following Clausius Clapeyron. Combining and solving for
-   ! frac yields:
-   
-   A=latent_heat_vapor*qsl*dt*inv_cp/(rv*T_endstep*T_endstep) &
-        *(latent_heat_sublim*qi2qv_sublim_tend + latent_heat_vapor*qr2qv_evap_tend)
-
-   frac = (qsl-qv+qv_sinks*dt + A)/(qv_sources*dt + A)
-
-   !The only way frac<0 is if qv-qv_sinks*dt is already greater than qsl. In this case
-   !the best we can do is zero out qv_sources.
-   frac = max(0._rtype,frac)
-
-   !The only way frac>1 is if qv-qv_sinks*dt+qv_sources*dt < qsl, in which case we shouldn't 
-   !limit anyways so set frac to 1:
-   frac = min(1._rtype,frac)
-   
-   qi2qv_sublim_tend = frac*qi2qv_sublim_tend
-   qr2qv_evap_tend = frac*qr2qv_evap_tend
+  real(rtype), intent(in) :: pres
+  real(rtype), intent(in) :: t_atm
+  real(rtype), intent(in) :: qv
+  real(rtype), intent(in) :: latent_heat_sublim,latent_heat_vapor
+  real(rtype), intent(in) :: dt
+  real(rtype), intent(in) :: qidep,qinuc
+  real(rtype), intent(inout) :: qi2qv_sublim_tend,qr2qv_evap_tend
+
+  real(rtype) :: qv_sinks, qv_sources, qv_endstep, T_endstep, qsl, A, frac
+
+  qv_sources = qi2qv_sublim_tend + qr2qv_evap_tend
+  if (qv_sources == 0._rtype) then
```

Comment:
@PeterCaldwell 

It's a bit risky to equate to a floating point (12 digit) number.
A condition is a better practice (or equate to an integer conditional switch).

Reply:
Yeah, testing caught this issue too. Fixed.